amazon-kinesis-client/src/test/java/software/amazon/kinesis/lifecycle/ShardConsumerTest.java

```diff
@@ -441,6 +454,131 @@ public final void testInitializationStateUponFailure() throws Exception {
         verify(initialState, never()).shutdownTransition(any());
     }
 
+    /**
+     * Test method to verify consumer undergoes the transition WAITING_ON_PARENT_SHARDS -> INITIALIZING -> PROCESSING
+     */
+    @SuppressWarnings("unchecked")
+    @Test
+    public final void testSuccessfulConsumerStateTransition() throws Exception {
+        ShardConsumer consumer = new ShardConsumer(recordsPublisher, executorService, shardInfo,
+                logWarningForTaskAfterMillis, shardConsumerArgument, blockedOnParentsState,
+                t -> t, 1, taskExecutionListener, 0);
+
+        mockSuccessfulUnblockOnParents();
+        mockSuccessfulInitializeWithFailureTransition();
+        mockSuccessfulProcessing(null);
+
+        int arbitraryExecutionCount = 3;
+        do {
+            try {
+                consumer.executeLifecycle();
+                Thread.sleep(100);
+            } catch (Exception e) {
+                // Suppress any exception like the scheduler.
+            }
+        } while (--arbitraryExecutionCount > 0);
+        assertEquals(ShardConsumerState.PROCESSING.consumerState().state(), consumer.currentState().state());
+    }
+
+    /**
+     * Test method to verify consumer does not transition to PROCESSING from WAITING_ON_PARENT_SHARDS when
+     * INITIALIZING tasks gets rejected.
+     */
+    @SuppressWarnings("unchecked")
+    @Test
+    public final void testConsumerNotTransitionsToProcessingWhenInitializationFails() {
+        ExecutorService failingService = spy(executorService);
+        ShardConsumer consumer = new ShardConsumer(recordsPublisher, failingService, shardInfo,
+                logWarningForTaskAfterMillis, shardConsumerArgument, blockedOnParentsState,
+                t -> t, 1, taskExecutionListener, 0);
+
+        mockSuccessfulUnblockOnParents();
+        mockSuccessfulInitializeWithFailureTransition();
+        mockSuccessfulProcessing(null);
+
+        // Failing the initialization task and all other attempts after that.
+        doCallRealMethod()
+                .doThrow(new RejectedExecutionException())
+                .when(failingService).execute(any());
+
+        int arbitraryExecutionCount = 5;
+        do {
+            try {
+                consumer.executeLifecycle();
+                Thread.sleep(100);
+            } catch (Exception e) {
+                // Suppress any exception like the scheduler.
+            }
+        } while (--arbitraryExecutionCount > 0);
+        assertEquals(ShardConsumerState.INITIALIZING.consumerState().state(), consumer.currentState().state());
```

Comment:
You may want to validate the number of calls to the executorservice.

Reply:
Added the expectation on number of calls